fedimint-core/src/config.rs

```diff
@@ -231,29 +228,42 @@ impl ClientConfig {
 ///
 /// The same `ModuleKind` may have multiple instances with different settings
 #[derive(Debug, Clone, Default)]
-pub struct ConfigGenParams(BTreeMap<String, serde_json::Value>);
```

Comment:
The idea here was that in the UI configuration we might not always have all the available fields (because there will be module we didn't even know about) etc. exposed, so we should have some form of free-form

```
some_key1: some_json_value1,
some_key2: some_json_value2,
...
```

configuration that user would be able to provide per-module.

So `ConfigGenParams` was already per-module-instance and `String` was just one key that user might give to the module.


`registry` really doesn't belong here, and then having the module look up it's own config is very undesireable.

Reply:
The `IModuleGen::trusted_dealer_gen` can take `ConfigGenParams` as `BTreeMap<String, serde_json::Value>`, but `ModuleGen::trusted_dealer_gen` (static typed version) could have `type ConfigGenParams` and then:

```
#[async_trait]
impl<T> IModuleGen for T
where
    T: ModuleGen + 'static + Sync,
{
```

would cast the untyped `BTreeMap<String, serde_json::Value>` to typed `ModuleGen::ConfigGenParams`, hiding the typed-erased json from the module author.

Reply:
I thought `ModuleInstanceId` identified each module instance, not a `String`?

Reply:
That map here was a map of 0 or more values, keyed on arbitrary string. So e.g. if someone added a `FederatedFilesystemModule` in the future, our UI would not know about it, so we would have no nice configuration boxes for it, but yet the user could add configuration manually for that particular module instance they are configuring in the config/UI and type in:

```
storage: "aws"
aws_credentials:
   token_id: ASDFADFADSF
   token_secret: j8j9sjfojasdodfjksadjf
```

and the whole thing would be put into the `ConfigGenParams` as a "map" and passed as a value to the `FederatedFilesystemModuleGen::distributed_gen` for that module.

The `Sting` in that map is a key: `storage` and `aws_credentials` in this case. We could make it into one big `serde_json::Value`, I supposed, but we never really implemented any of this, and it seems it confused everyone that looked at it.

Reply:
So it sounds like what we need is something to store `ModuleInstanceId` -> `ConfigGenParams`?

And then each module instance gets its own `ConfigGenParams`?  Not sure that `ConfigGenParams` needs to be a map, it could just be a `JsonWithKind`.

Reply:
> So it sounds like what we need is something to store `ModuleInstanceId` -> `ConfigGenParams`?

Possibly, yes.

> And then each module instance gets its own `ConfigGenParams`? 

Yes.

> Not sure that `ConfigGenParams` needs to be a map, it could just be a `JsonWithKind`.

Possibly, yes.  There are some considerations. `serde_json::Value` possibly being an array, null etc, which makes less sense, but whatever. It's not really used for anything yet, so we can play with it once we do.